core/src/test/java/org/apache/iceberg/catalog/CatalogTransactionTests.java

```diff
@@ -0,0 +1,736 @@
+    Table two = catalog().loadTable(second);
+
+    CatalogTransaction catalogTransaction = catalog().createTransaction(isolationLevel);
+    Catalog txCatalog = catalogTransaction.asCatalog();
+    txCatalog.loadTable(first).newAppend().appendFile(FILE_A).appendFile(FILE_D).commit();
+    assertThat(Iterables.size(txCatalog.loadTable(first).newScan().planFiles())).isEqualTo(2);
+    assertThat(Iterables.size(txCatalog.loadTable(second).newScan().planFiles())).isEqualTo(0);
+
+    two.newFastAppend().appendFile(FILE_B).appendFile(FILE_C).appendFile(FILE_D).commit();
+
+    // this should not be allowed with SERIALIZABLE after the table has been already read
+    // within the catalog TX, but is allowed with SNAPSHOT
+    // catalog TX should still the version of the table it initially read (with 0 files)
+    assertThat(Iterables.size(txCatalog.loadTable(second).newScan().planFiles())).isEqualTo(0);
+
+    if (SERIALIZABLE == isolationLevel) {
+      Assertions.assertThatThrownBy(catalogTransaction::commitTransaction)
+          .isInstanceOf(ValidationException.class)
+          .hasMessage(
+              "SERIALIZABLE isolation violation: Found table metadata updates to table 'ns.b' after it was read");
+
+      assertThat(Iterables.size(catalog().loadTable(first).newScan().planFiles())).isEqualTo(0);
+      assertThat(Iterables.size(catalog().loadTable(second).newScan().planFiles())).isEqualTo(3);
+    } else {
+      catalogTransaction.commitTransaction();
+
+      assertThat(Iterables.size(catalog().loadTable(first).newScan().planFiles())).isEqualTo(2);
+      assertThat(Iterables.size(catalog().loadTable(second).newScan().planFiles())).isEqualTo(3);
+    }
+  }
+
+  @Test
+  public void concurrentTx() {
+    concurrentTx(SNAPSHOT);
+  }
+
+  @Test
+  public void concurrentTxWithSerializable() {
+    concurrentTx(SERIALIZABLE);
+  }
+
+  private void concurrentTx(CatalogTransaction.IsolationLevel isolationLevel) {
+    TableIdentifier identifier = TableIdentifier.of("ns", "tbl");
+    catalog().createTable(identifier, SCHEMA);
+    Table one = catalog().loadTable(identifier);
+
+    CatalogTransaction catalogTransaction = catalog().createTransaction(isolationLevel);
+    Catalog txCatalog = catalogTransaction.asCatalog();
+
+    // perform updates outside catalog TX but before table has been read inside the catalog TX
+    one.newAppend().appendFile(FILE_A).appendFile(FILE_D).commit();
+
+    Snapshot snapshot = ((BaseTable) one).operations().refresh().currentSnapshot();
+    assertThat(snapshot).isNotNull();
+    assertThat(snapshot.summary().get(SnapshotSummary.ADDED_FILES_PROP)).isEqualTo("2");
+    assertThat(snapshot.summary().get(SnapshotSummary.ADDED_RECORDS_PROP)).isEqualTo("2");
+    assertThat(snapshot.summary().get(SnapshotSummary.TOTAL_DATA_FILES_PROP)).isEqualTo("2");
+
+    // this should not fail with any isolation level
+    txCatalog.loadTable(identifier).newAppend().appendFile(FILE_A).appendFile(FILE_B).commit();
```

Comment:
It would be helpful to use names that tie tables and identifiers together, like `oneIdent` so that it is clear that this is loading table `one`.

Also, this adds the same file to the table twice, which is probably not a good thing to do in tests because it doesn't match real-world behavior.